### PR TITLE
Middlewares

### LINKED ISSUE #770: Middlewares
Related to the discussion in https://github.com/prisma/prisma-client-js/issues/669

Users of Prisma Client need a way to hook into Prisma Client.
The system that allows this hooking-in needs to allow the following use-cases:

- Monitoring, measuring performance, logging
- Overwriting args
- Validation

When having hooks in a system like Prisma Client, we can differentiate between blocking and non-blocking hooks. Blocking would mean - this function needs to succeed before we continue with executing the Prisma Client query. Non-blocking would be a traditional event emitter, for example used for logging. It will be called async and doesn't have to succeed for the client to continue with its business.

**Why am I calling this a middleware instead of a hook?**
A middleware clearly states, that it is the "blocking thing" that can impact the control flow of what Prisma Client does.

As a first start, we should think about a simple but effective API, that solves as many use-cases as possible, while minimizing the API surface.

The API I suggest for that looks like this:

```ts
prisma.use(async ({ fetchQuery, params }) => {
  console.log('Before')
  const data = await fetchQuery(params)
  console.log('After')
})
```

This hook or **middleware** will then be called before every query, for example `prisma.user.findMany()`.
The params include all the relevant information, like the args, the model and the method being used.
Details of the params have to be figured out.

With this approach, we can cover all the above-mentioned use-cases.

Please comment, if this API would be useful for you and if any use-cases are left out by this.

### PR DESCRIPTION
**Pull request by timsuchanek**
Closes https://github.com/prisma/prisma-client-js/issues/770

### CODE REVIEW

File: src/packages/client/src/runtime/utils/printStack.ts [hunk 1]
```diff
@@ -51,15 +51,17 @@ export const printStack = ({
     const stack = stackTraceParser.parse(callsite)
     // TODO: more resilient logic to check that it's not relative to cwd
     const trace = stack.find((t, i) => {
-      if (i < 2) {
+      if (i < 3) {
```
Comment (Jolg42) [inline src/packages/client/src/runtime/utils/printStack.ts:54]: Looks like a magic number to me 😄

Comment (timsuchanek) [reply src/packages/client/src/runtime/utils/printStack.ts:54]: It's about how deep the stack can be cut away - we have one more function call in our stack, so we can ignore that level. I'll add a comment.

Comment (Jolg42) [reply src/packages/client/src/runtime/utils/printStack.ts:54]: Thanks!

File: src/packages/client/src/runtime/getPrismaClient.ts [hunk 2]
```diff
@@ -595,6 +621,152 @@ export function getPrismaClient(config: GetPrismaClientOptions): any {
...
+        document,
+        clientMethod,
+        typeName,
+        dataPath,
+        isList,
+        rootField: rootField!,
+        callsite,
+        showColors: this.errorFormat === 'pretty',
+        args,
+        engineHook: this._engineMiddlewares[0],
```
Comment (Jolg42) [inline src/packages/client/src/runtime/getPrismaClient.ts:765]: Should engineHook be renamed engineMiddleware?

File: src/packages/client/src/__tests__/runtime-tests/middlewares/index.js [hunk 3]
```diff
@@ -7,13 +7,24 @@ module.exports = async () => {
   const allResults = []
   const engineResults = []
```
Comment (Jolg42) [inline src/packages/client/src/__tests__/runtime-tests/middlewares/index.js:9]: It's missing a `db.use('all'` I think (The default is all though.)

**Review by Jolg42 — APPROVED** [review]

File: src/packages/client/src/__tests__/runtime-tests/middlewares/index.js (hunk 3 above)
Comment (timsuchanek) [reply src/packages/client/src/__tests__/runtime-tests/middlewares/index.js:9]: We don't want to document `db.use('all'` yet, hence no test yet.

File: src/packages/client/src/runtime/getPrismaClient.ts (hunk 2 above)
Comment (timsuchanek) [reply src/packages/client/src/runtime/getPrismaClient.ts:765]: The engines part is an internal api for now that we don't want to document.
I want to expose that properly once we need it.

**Pull request merged**